File: src/go/control_command.go
package PhantasyIslandGoRemoteControl

import (
	"fmt"
)

// AirplaneController 无人机控制
// 嵌入了 AirplaneCore 以实现继承效果
type AirplaneController struct {
	*AirplaneCore
	Count      int
	IsFastMode bool
}

func NewAirplaneController(core *AirplaneCore) *AirplaneController {
	return &AirplaneController{
		AirplaneCore: core,
		Count:        1,
		IsFastMode:   false,
	}
}

// UseFastMode 设置命令是否使用非阻塞模式
// Go 中实现 Future 模式通常使用 channel 或直接开启 goroutine。
// 这里简化实现，主要区分同步和异步调用。
func (ac *AirplaneController) UseFastMode(fastMode bool) {
	ac.IsFastMode = fastMode
}

func (ac *AirplaneController) nextCount() int {
	ac.Count += 2
	return ac.Count
}

func (ac *AirplaneController) prepareCommand(command string) string {
	return fmt.Sprintf("%s %d %s", ac.KeyName, ac.nextCount(), command)
}

func (ac *AirplaneController) sendCmd(command string) interface{} {
	cmdStr := ac.prepareCommand(command)
	if ac.IsFastMode {
		return SendCmdVolatile(cmdStr)
	}
	return SendCmd(cmdStr)
}

func (ac *AirplaneController) Mode(mode int) interface{} {
	return ac.AirplaneMode(mode)
}

func (ac *AirplaneController) Takeoff(high int) interface{} {
	return ac.sendCmd(fmt.Sprintf("takeoff %d", high))
}

func (ac *AirplaneController) Land() interface{} {
	return ac.sendCmd("land")
}

func (ac *AirplaneController) Emergency() interface{} {
	return ac.sendCmd("emergency")
}

func (ac *AirplaneController) Up(distance int) interface{} {
	return ac.sendCmd(fmt.Sprintf("up %d", distance))
}

func (ac *AirplaneController) Down(distance int) interface{} {
	return ac.sendCmd(fmt.Sprintf("down %d", distance))
}

func (ac *AirplaneController) Forward(distance int) interface{} {
	return ac.sendCmd(fmt.Sprintf("forward %d", distance))
}

func (ac *AirplaneController) Back(distance int) interface{} {
	return ac.sendCmd(fmt.Sprintf("back %d", distance))
}

func (ac *AirplaneController) Left(distance int) interface{} {
	return ac.sendCmd(fmt.Sprintf("left %d", distance))
}

func (ac *AirplaneController) Right(distance int) interface{} {
	return ac.sendCmd(fmt.Sprintf("right %d", distance))
}

func (ac *AirplaneController) Goto(x int, y int, h int) interface{} {
	return ac.sendCmd(fmt.Sprintf("goto %d %d %d", x, y, h))
}

func (ac *AirplaneController) Flip(direction string) interface{} {
	return ac.sendCmd(fmt.Sprintf("flip %s 1", direction))
}

func (ac *AirplaneController) FlipForward() interface{} {
	return ac.Flip("f")
}

func (ac *AirplaneController) FlipBack() interface{} {
	return ac.Flip("b")
}

func (ac *AirplaneController) FlipLeft() interface{} {
	return ac.Flip("l")
}

func (ac *AirplaneController) FlipRight() interface{} {
	return ac.Flip("r")
}

func (ac *AirplaneController) Rotate(degree int) interface{} {
	return ac.sendCmd(fmt.Sprintf("rotate %d", degree))
}

func (ac *AirplaneController) Cw(degree int) interface{} {
	return ac.sendCmd(fmt.Sprintf("cw %d", degree))
}

func (ac *AirplaneController) Ccw(degree int) interface{} {
	return ac.sendCmd(fmt.Sprintf("ccw %d", degree))
}

func (ac *AirplaneController) High(high int) interface{} {
	return ac.sendCmd(fmt.Sprintf("high %d", high))
}

func (ac *AirplaneController) Speed(speed int) interface{} {
	return ac.sendCmd(fmt.Sprintf("setSpeed %d", speed))
}

func (ac *AirplaneController) Led(r int, g int, b int) interface{} {
	return ac.sendCmd(fmt.Sprintf("light %d %d %d", r, g, b))
}

func (ac *AirplaneController) Bln(r int, g int, b int) interface{} {
	return ac.sendCmd(fmt.Sprintf("bln %d %d %d", r, g, b))
}

func (ac *AirplaneController) Rainbow(r int, g int, b int) interface{} {
	return ac.sendCmd(fmt.Sprintf("rainbow %d %d %d", r, g, b))
}

func (ac *AirplaneController) AirplaneMode(mode int) interface{} {
	return ac.sendCmd(fmt.Sprintf("airplane_mode %d", mode))
}

func (ac *AirplaneController) Stop() interface{} {
	return ac.Hover()
}

func (ac *AirplaneController) Hover() interface{} {
	return ac.sendCmd("hover")
}

File: src/rs/src/control_command.rs
use crate::airplane_core::AirplaneCore;
use crate::http_layer::{send_cmd, send_cmd_volatile};
use serde_json::Value;

pub struct AirplaneController {
    pub core: AirplaneCore,
    pub count: i32,
    pub fast_mode: bool,
}

impl AirplaneController {
    pub fn new(
        key_name: String,
        type_name: String,
        update_timestamp: i64,
        status: crate::airplane_core::AirplaneFlyStatus,
        camera_front: String,
        camera_down: String,
    ) -> Self {
        Self {
            core: AirplaneCore::new(
                key_name,
                type_name,
                update_timestamp,
                status,
                camera_front,
                camera_down,
            ),
            count: 1,
            fast_mode: false,
        }
    }

    pub fn use_fast_mode(&mut self, fast_mode: bool) {
        self.fast_mode = fast_mode;
    }

    fn next_count(&mut self) -> i32 {
        self.count += 2;
        self.count
    }

    fn prepare_command(&mut self, command: &str) -> String {
        let count = self.next_count();
        format!("{} {} {}", self.core.key_name, count, command)
    }

    fn send_cmd(&mut self, command: &str) -> Value {
        let full_cmd = self.prepare_command(command);
        if self.fast_mode {
            send_cmd_volatile(&full_cmd)
        } else {
            send_cmd(&full_cmd)
        }
    }

    pub fn mode(&mut self, mode: i32) -> Value {
        self.airplane_mode(mode)
    }

    pub fn takeoff(&mut self, high: i32) -> Value {
        self.send_cmd(&format!("takeoff {}", high))
    }

    pub fn land(&mut self) -> Value {
        self.send_cmd("land")
    }

    pub fn emergency(&mut self) -> Value {
        self.send_cmd("emergency")
    }

    pub fn up(&mut self, distance: i32) -> Value {
        self.send_cmd(&format!("up {}", distance))
    }

    pub fn down(&mut self, distance: i32) -> Value {
        self.send_cmd(&format!("down {}", distance))
    }

    pub fn forward(&mut self, distance: i32) -> Value {
        self.send_cmd(&format!("forward {}", distance))
    }

    pub fn back(&mut self, distance: i32) -> Value {
        self.send_cmd(&format!("back {}", distance))
    }

    pub fn left(&mut self, distance: i32) -> Value {
        self.send_cmd(&format!("left {}", distance))
    }

    pub fn right(&mut self, distance: i32) -> Value {
        self.send_cmd(&format!("right {}", distance))
    }

    pub fn goto(&mut self, x: i32, y: i32, h: i32) -> Value {
        self.send_cmd(&format!("goto {} {} {}", x, y, h))
    }

    pub fn flip(&mut self, direction: &str) -> Value {
        self.send_cmd(&format!("flip {} 1", direction))
    }

    pub fn rotate(&mut self, degree: i32) -> Value {
        self.send_cmd(&format!("rotate {}", degree))
    }

    pub fn cw(&mut self, degree: i32) -> Value {
        self.send_cmd(&format!("cw {}", degree))
    }

    pub fn ccw(&mut self, degree: i32) -> Value {
        self.send_cmd(&format!("ccw {}", degree))
    }

    pub fn high(&mut self, high: i32) -> Value {
        self.send_cmd(&format!("high {}", high))
    }

    pub fn speed(&mut self, speed: i32) -> Value {
        self.send_cmd(&format!("setSpeed {}", speed))
    }

    pub fn led(&mut self, r: i32, g: i32, b: i32) -> Value {
        self.send_cmd(&format!("light {} {} {}", r, g, b))
    }

    pub fn bln(&mut self, r: i32, g: i32, b: i32) -> Value {
        self.send_cmd(&format!("bln {} {} {}", r, g, b))
    }

    pub fn rainbow(&mut self, r: i32, g: i32, b: i32) -> Value {
        self.send_cmd(&format!("rainbow {} {} {}", r, g, b))
    }

    pub fn airplane_mode(&mut self, mode: i32) -> Value {
        self.send_cmd(&format!("airplane_mode {}", mode))
    }

    pub fn hover(&mut self) -> Value {
        self.send_cmd("hover")
    }

    pub fn stop(&mut self) -> Value {
        self.hover()
    }
}
